File: kernel/drivers/cc/tty.lua
local BaseTty = require("drivers.common.tty")

-- ANSI colour code → CC colours constant
local ANSI_FG = {
    [30] = colours.black,  [31] = colours.red,    [32] = colours.green,
    [33] = colours.yellow, [34] = colours.blue,   [35] = colours.purple,
    [36] = colours.cyan,   [37] = colours.white,
}
local ANSI_BG = {
    [40] = colours.black,  [41] = colours.red,    [42] = colours.green,
    [43] = colours.yellow, [44] = colours.blue,   [45] = colours.purple,
    [46] = colours.cyan,   [47] = colours.white,
}

---@class CCTty: BaseTty
local CCTty = BaseTty:new()

function CCTty:_write_raw(ch)
    io.write(ch)
end

function CCTty:_echo_erase(n)
    local x, y = term.getCursorPos()
    term.setCursorPos(x - n, y)
    term.write(string.rep(" ", n))
    term.setCursorPos(x - n, y)
end

function CCTty:_set_cursor_pos(x, y)
    term.setCursorPos(x, y)
end

function CCTty:_get_cursor_pos()
    return term.getCursorPos()
end

function CCTty:_get_term_size()
    return term.getSize()
end

function CCTty:_clear_screen()
    term.clear()
    term.setCursorPos(1, 1)
end

function CCTty:_clear_to_eol()
    local x, y = term.getCursorPos()
    local w, _ = term.getSize()
    term.write(string.rep(" ", w - x + 1))
    term.setCursorPos(x, y)
end

function CCTty:_clear_line()
    term.clearLine()
    local _, y = term.getCursorPos()
    term.setCursorPos(1, y)
end

function CCTty:_set_text_color(ansi)
    if ANSI_FG[ansi] then term.setTextColour(ANSI_FG[ansi]) end
end

function CCTty:_set_bg_color(ansi)
    if ANSI_BG[ansi] then term.setBackgroundColour(ANSI_BG[ansi]) end
end

function CCTty:_reset_colors()
    term.setTextColour(colours.white)
    term.setBackgroundColour(colours.black)
end

function CCTty:_set_cursor_visible(visible)
    term.setCursorBlink(visible)
end

return CCTty

File: kernel/drivers/cc/tty.ts
import * as expect from "cc/expect";
import { BaseTty } from "../common/tty/init";

let ANSI_FG = new Map<integer, integer>([
	[30, colors.black],
	[31, colors.red],
	[32, colors.green],
	[33, colors.yellow],
	[34, colors.blue],
	[35, colors.purple],
	[36, colors.cyan],
	[37, colors.white],
]);

let ANSI_BG = new Map<integer, integer>([
	[40, colors.black],
	[41, colors.red],
	[42, colors.green],
	[43, colors.yellow],
	[44, colors.blue],
	[45, colors.purple],
	[46, colors.cyan],
	[47, colors.white],
]);

export class CCTty extends BaseTty {
	write_raw(char: char) {
		io.write(char);
	}

	echo_erase(n: integer) {
		let [x, y] = term.getCursorPos();
		term.setCursorPos(x - n, y);
		term.write(string.rep(" ", n));
		term.setCursorPos(x - n, y);
	}

	set_cursor_pos(x: integer, y: integer) {
		term.setCursorPos(x, y);
	}

	get_cursor_pos() {
		return term.getCursorPos();
	}

	get_term_size(): [integer, integer] {
		return term.getSize();
	}

	clear_screen(): void {
		term.clear();
		term.setCursorPos(1, 1);
	}

	clear_to_eol() {
		let [x, y] = term.getCursorPos();
		let [w, _] = term.getSize();
		term.write(string.rep(" ", w - x + 1));
		term.setCursorPos(x, y);
	}

	clear_line() {
		term.clearLine();
		let [_, y] = term.getCursorPos();
		term.setCursorPos(1, y);
	}

	set_text_color(ansi: integer) {
		let code = ANSI_FG.get(ansi);
		if (code !== undefined) {
			term.setTextColor(code);
		}
	}

	set_bg_color(ansi: integer) {
		let code = ANSI_BG.get(ansi);
		if (code !== undefined) {
			term.setBackgroundColor(code);
		}
	}

	reset_colors(): void {
		term.setTextColor(colors.white);
		term.setBackgroundColor(colors.black);
	}

	set_cursor_visible(visible: boolean): void {
		term.setCursorBlink(visible);
	}
}
